core/src/main/scala/org/http4s/internal/Random.scala

```diff
@@ -0,0 +1,110 @@
+private[http4s] trait Random[F[_]] {
+
+  /** Generates a random BigInt between 0 and (2^bits-1) inclusive. */
+  def nextBigInt(bits: Int): F[BigInt]
+}
+
+private[http4s] object Random {
+
+  /** Create a non-blocking Random instance.  This is more efficient
+    * than [[javaSecurityRandom]], but requires a more thoughtful
+    * instance of random
+    *
+    * @param random a non-blocking instance of java.util.Random.  It
+    * is the caller's responsibility to assure that the instance is
+    * non-blocking.  If the instance blocks during seeding, it is the
+    * caller's responsibility to seed it.
+    */
+  def javaUtilRandomNonBlocking[F[_]](random: JRandom)(implicit F: Sync[F]): Random[F] =
+    new Random[F] {
+      def nextBigInt(bits: Int): F[BigInt] =
+        F.delay(BigInt(new JBigInteger(bits, random)))
+    }
+
+  /** Creates a blocking Random instance.  All calls to nextBytes are
+    * shifted to the blocking pool.  This is safer, but less
+    * effecient, than [[javaUtilRandomNonBlocking]].
+    *
+    * @param random a potentially blocking instance of java.util.Random
+    */
+  def javaUtilRandomBlocking[F[_]](
+      blocker: Blocker
+  )(random: JRandom)(implicit F: Sync[F], cs: ContextShift[F]): Random[F] =
+    new Random[F] {
+      def nextBigInt(bits: Int): F[BigInt] =
+        blocker.blockOn(F.delay(BigInt(new JBigInteger(bits, random))))
+    }
+
+  /** Creates a Random instance.  On most platforms, it will be
+    * non-blocking.  If a non-blocking instance can't be guaranteed,
+    * falls back to a blocking implementation.
+    */
+  def javaSecuritySecureRandom[F[_]](blocker: Blocker)(implicit F: Sync[F], cs: ContextShift[F]) = {
+    // This is a known, non-blocking, threadsafe algorithm
+    def happyRandom = F.delay(SecureRandom.getInstance("NativePRNGNonBlocking"))
+
+    // If we can't sniff out a more optimal solution, we can always
+    // fall back to a pool of blocking instances
+    def fallbackPool =
+      F.delay(Runtime.getRuntime.availableProcessors())
+        .flatMap(pool(_)(F.delay(new SecureRandom()).map(javaUtilRandomBlocking(blocker))))
+
+    javaMajorVersion.flatMap {
+      case Some(major) if major > 8 =>
+        happyRandom.redeemWith(
+          // We can't guarantee that our SecureRandom is both
+          // non-blocking and threadsafe.
+          _ => fallbackPool,
+          // We are thread safe and non-blocking.  This is the
+          // happy path, and happily, the common path.
+          rnd => F.pure(javaUtilRandomNonBlocking(rnd)),
+        )
+
+      case Some(_) | None =>
+        // We can't guarantee we're not stuck in a mutex.
+        fallbackPool
+    }
+  }
+
+  def pool[F[_]](n: Int)(random: F[Random[F]])(implicit F: Sync[F]): F[Random[F]] =
+    for {
+      ref <- Ref[F].of(0)
+      array <- Vector.fill(n)(random).sequence
+    } yield {
+      val incrGet = ref.modify(i => (if (i < (n - 1)) i + 1 else 0, i))
```

Comment:
This is just `(i + 1) % n`, right? 😅

Reply:
Uh, probably.  I stole it from CE3 Random and neglected to attribute it.